azure-iot-device/tests/common/pipeline/test_pipeline_ops_base.py

```diff
@@ -7,64 +7,268 @@
+pytestmark = pytest.mark.usefixtures("fake_pipeline_thread")
 
 
-@pytest.mark.describe("PipelineOperation")
-class TestPipelineOperation(object):
-    @pytest.mark.it("Can't be instantiated")
-    def test_instantiate(self, mocker):
-        with pytest.raises(TypeError):
-            pipeline_ops_base.PipelineOperation(mocker.MagicMock())
+class ConnectOperationTestConfig(object):
+    @pytest.fixture
+    def cls_type(self):
+        return pipeline_ops_base.ConnectOperation
 
+    @pytest.fixture
+    def init_kwargs(self, mocker):
+        kwargs = {"callback": mocker.MagicMock()}
+        return kwargs
 
-pipeline_data_object_test.add_operation_test(
-    cls=pipeline_ops_base.ConnectOperation, module=this_module
+
+pipeline_ops_test.add_operation_tests(
+    test_module=this_module,
+    op_class_under_test=pipeline_ops_base.ConnectOperation,
+    op_test_config_class=ConnectOperationTestConfig,
 )
-pipeline_data_object_test.add_operation_test(
-    cls=pipeline_ops_base.DisconnectOperation, module=this_module
+
+
+class DisconnectOperationTestConfig(object):
+    @pytest.fixture
+    def cls_type(self):
+        return pipeline_ops_base.DisconnectOperation
+
+    @pytest.fixture
+    def init_kwargs(self, mocker):
+        kwargs = {"callback": mocker.MagicMock()}
+        return kwargs
+
+
+pipeline_ops_test.add_operation_tests(
+    test_module=this_module,
+    op_class_under_test=pipeline_ops_base.DisconnectOperation,
+    op_test_config_class=DisconnectOperationTestConfig,
 )
-pipeline_data_object_test.add_operation_test(
-    cls=pipeline_ops_base.ReconnectOperation, module=this_module
+
+
+class ReconnectOperationTestConfig(object):
+    @pytest.fixture
+    def cls_type(self):
+        return pipeline_ops_base.ReconnectOperation
+
+    @pytest.fixture
+    def init_kwargs(self, mocker):
+        kwargs = {"callback": mocker.MagicMock()}
+        return kwargs
+
+
+pipeline_ops_test.add_operation_tests(
+    test_module=this_module,
+    op_class_under_test=pipeline_ops_base.ReconnectOperation,
+    op_test_config_class=ReconnectOperationTestConfig,
 )
-pipeline_data_object_test.add_operation_test(
-    cls=pipeline_ops_base.EnableFeatureOperation,
-    module=this_module,
-    positional_arguments=["feature_name", "callback"],
+
+
+class EnableFeatureOperationTestConfig(object):
+    @pytest.fixture
+    def cls_type(self):
+        return pipeline_ops_base.EnableFeatureOperation
+
+    @pytest.fixture
+    def init_kwargs(self, mocker):
+        kwargs = {"feature_name": "some_feature", "callback": mocker.MagicMock()}
+        return kwargs
+
+
+class EnableFeatureInstantiationTests(EnableFeatureOperationTestConfig):
+    @pytest.mark.it(
+        "Initializes 'feature_name' attribute with the provided 'feature_name' parameter"
+    )
+    def test_feature_name(self, cls_type, init_kwargs):
+        op = cls_type(**init_kwargs)
+        assert op.feature_name == init_kwargs["feature_name"]
```

Comment:
What's the benefit of writing these tests by hand rather than using kwargs to parametrize them?

Reply:
As discussed offline, it allows for simpler, more readable tests. I don't mind using a function to add a series of test classes dynamically into a test module, but only when those tests are still statically laid out in a common place. When the tests themselves are dynamically created, I feel it becomes harder to follow (especially for someone not familiar with the codebase)

---
In reply to: [346512691](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r346512691) [](ancestors = 346512691)